crates/subspace-farmer/src/plot.rs

```diff
@@ -443,37 +443,55 @@ impl WeakPlot {
     }
 }
 
+/// Mapping from piece index hash to piece offset.
+///
+/// Piece index hashes are transformed in the following manner:
+/// - Assume that farmer address is the middle (`2 ^ 255`) of the `PieceDistance` field
+/// - Move every piece according to that
 #[derive(Debug)]
 struct IndexHashToOffsetDB {
     inner: DB,
     address: PublicKey,
-    max_distance: Option<PieceDistance>,
+    /// The first one is the lower bound of plot, the second one is the upper bound
+    key_bounds: Option<(PieceDistance, PieceDistance)>,
 }
 
 impl IndexHashToOffsetDB {
-    fn update_max_distance(&mut self) {
-        self.max_distance = {
+    fn update_key_bounds(&mut self) {
+        self.key_bounds = try {
             let mut iter = self.inner.raw_iterator();
+            iter.seek_to_first();
+            let lower_bound = iter.key().map(PieceDistance::from_big_endian)?;
             iter.seek_to_last();
-            iter.key().map(PieceDistance::from_big_endian)
+            let upper_bound = iter.key().map(PieceDistance::from_big_endian)?;
+            (lower_bound, upper_bound)
         };
     }
 
+    fn get_key(&self, index_hash: &PieceIndexHash) -> PieceDistance {
+        let address = PieceDistance::from_big_endian(self.address.as_ref());
+        let index_hash = PieceDistance::from_big_endian(&index_hash.0);
+
+        // We do not care about overflows here
+        (PieceDistance::MAX / 2u64)
+            .overflowing_add(index_hash.overflowing_sub(address).0)
+            .0
+    }
+
     fn open_default(path: impl AsRef<Path>, address: PublicKey) -> Result<Self, PlotError> {
```

Comment:
Let's keep `open` the first method. Having first method one that creates instance and last that destroys (if any) maps nicely to the lifecycle order of the entity.